src/unicode/mod.rs

```diff
@@ -1,50 +1,122 @@
+            self.0.as_ref().chars().flat_map(lookup)
+                .zip(other.0.as_ref().chars().flat_map(lookup))
+                .all(|(a, b)| a == b)
+        }
     }
-}
 
-impl<S: AsRef<str>> Eq for Unicode<S> {}
+    #[cfg(__unicase__iter_cmp)]
+    impl<T: AsRef<str>> Ord for Unicode<T> {
+        #[inline]
+        fn cmp(&self, other: &Self) -> Ordering {
+            let self_chars = self.0.as_ref().chars().flat_map(lookup);
+            let other_chars = other.0.as_ref().chars().flat_map(lookup);
+            self_chars.cmp(other_chars)
+        }
+    }
 
-#[cfg(__unicase__iter_cmp)]
-impl<T: AsRef<str>> PartialOrd for Unicode<T> {
-    #[inline]
-    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
-        Some(self.cmp(other))
+    impl<S: AsRef<str>> Hash for Unicode<S> {
+        #[inline]
+        fn hash<H: Hasher>(&self, hasher: &mut H) {
+            let mut buf = [0; 4];
+            for c in self.0.as_ref().chars().flat_map(|c| lookup(c)) {
+                let len = char_to_utf8(c, &mut buf);
+                hasher.write(&buf[..len])
+            }
+        }
     }
 }
 
-#[cfg(__unicase__iter_cmp)]
-impl<T: AsRef<str>> Ord for Unicode<T> {
-    #[inline]
-    fn cmp(&self, other: &Self) -> Ordering {
-        let self_chars = self.0.as_ref().chars().flat_map(lookup);
-        let other_chars = other.0.as_ref().chars().flat_map(lookup);
-        self_chars.cmp(other_chars)
+#[cfg(target_arch = "wasm32")]
+mod wasm {
+    #[cfg(__unicase__iter_cmp)]
+    use std::cmp::Ordering;
+
+    use std::hash::{Hash, Hasher};
+
+    use js_sys::{JsString, Array, Object, Reflect};
+
+    use super::{Unicode, char_to_utf8};
+
+    impl<S1: AsRef<str>, S2: AsRef<str>> PartialEq<Unicode<S2>> for Unicode<S1> {
+        #[inline]
+        fn eq(&self, other: &Unicode<S2>) -> bool {
+            js_compare(self.0.as_ref(), other.0.as_ref()) == 0
+        }
+    }
+
+    #[cfg(__unicase__iter_cmp)]
+    impl<T: AsRef<str>> Ord for Unicode<T> {
+        #[inline]
+        fn cmp(&self, other: &Self) -> Ordering {
+            match js_compare(self.0.as_ref(), other.0.as_ref()) {
+                0 => Ordering::Equal,
+                i if i < 0 => Ordering::Less,
+                i if i > 0 => Ordering::Greater,
+                _ => unreachable!()
+            }
+        }
+    }
+
+    impl<S: AsRef<str>> Hash for Unicode<S> {
+        #[inline]
+        fn hash<H: Hasher>(&self, hasher: &mut H) {
+            let mut buf = [0; 4];
+            for c in js_to_lowercase(self.0.as_ref()).chars() {
+                let len = char_to_utf8(c, &mut buf);
+                hasher.write(&buf[..len])
+            }
+        }
+    }
+
+    fn js_compare(a: &str, b: &str) -> i32 {
+        let locales = Array::new();
+        locales.push(&"en".into());
+
+        let options = Object::new();
+        Reflect::set(&options, &"sensitivity".into(), &"accent".into()).unwrap();
+
+        let a: JsString = a.into();
```

Comment:
It's too bad it looks like this copies the full string into the JS heap, but there doesn't seem to be a better way...

Reply:
Unfortunately not yea, as far as I know. It should still be fairly efficient because it uses TextEncoder/TextDecoder rather than converting it manually IIRC, but there is an unnecessary copy back and forth.